scipy/ndimage/tests/test_filters.py

```diff
@@ -12,15 +13,102 @@
 from scipy import ndimage
 from scipy.ndimage.filters import _gaussian_kernel1d, rank_filter
 
-from . import types, float_types
+from . import types, float_types, complex_types
 
 
 def sumsq(a, b):
     return math.sqrt(((a - b)**2).sum())
 
 
+def _complex_correlate(array, kernel, real_dtype, convolve=False):
+    """Utility to perform a reference complex-valued convolutions.
+
+    When convolve==False, correlation is performed instead
+    """
+    array = numpy.asarray(array)
+    kernel = numpy.asarray(kernel)
+    complex_array = array.dtype.kind == 'c'
+    complex_kernel = kernel.dtype.kind == 'c'
+    if array.ndim == 1:
+        func = ndimage.convolve1d if convolve else ndimage.correlate1d
+    else:
+        func = ndimage.convolve if convolve else ndimage.correlate
+    if not convolve:
+        kernel = kernel.conj()
+    if complex_array and complex_kernel:
+        output = (
+            func(array.real, kernel.real, output=real_dtype) -
+            func(array.imag, kernel.imag, output=real_dtype) +
+            1j * func(array.imag, kernel.real, output=real_dtype) +
+            1j * func(array.real, kernel.imag, output=real_dtype)
+        )
+    elif complex_array:
+        output = (
+            func(array.real, kernel, output=real_dtype) +
+            1j * func(array.imag, kernel, output=real_dtype)
+        )
```

Comment:
Can we get a little bit of memory efficiency and no need for `+` via something like:
```
output = np.empty(array.shape, dtype=np.complex128).view(np.float64)
func(array.real, kernel, output=output[..., ::2])
func(array.imag, kernel, output=output[..., 1::2])
```

Reply:
I think so, but I left it that way by intention to try and make this test reference implmentation unambiguous. This version is only use to validate the true implementation in `scipy.ndimage.filters._complex_via_real_components`, so it only ever receives tiny images from the test cases. 

Currently it takes around 6 seconds on my machine to run the full ndimage test suite, so I don't think it is too critical to reduce the test times further.

I am interested in having the user-facing implementation within the `_complex_via_real_components` function within `filters.py` be efficient as possible. Within that version, I did do try to do this type of thing (providing `output=output.real`, etc).

Reply:
Whoops, I didn't even notice this was in `test_*` when I was reading this! Yes I agree it's preferable in that case to leave it as the current, clearer implementation